flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java

```diff
@@ -521,6 +529,49 @@ private static void validateImplementationMethod(
         }
     }
 
+    private static void validateAsyncImplementationMethod(
+            Class<? extends UserDefinedFunction> clazz, String... methodNameOptions) {
+        final Set<String> nameSet = new HashSet<>(Arrays.asList(methodNameOptions));
+        final List<Method> methods = getAllDeclaredMethods(clazz);
+        for (Method method : methods) {
+            if (!nameSet.contains(method.getName())) {
+                continue;
+            }
+            if (!method.getReturnType().equals(Void.TYPE)) {
+                throw new ValidationException(
+                        String.format(
+                                "Method '%s' of function class '%s' must be void.",
+                                method.getName(), clazz.getName()));
+            }
+            boolean foundParam = false;
+            boolean genericParam = false;
+            if (method.getParameterCount() >= 1) {
+                Type firstParam = method.getGenericParameterTypes()[0];
+                firstParam = ExtractionUtils.resolveVariableWithClassContext(clazz, firstParam);
+                if (CompletableFuture.class.equals(firstParam)) {
+                    foundParam = true;
+                } else if (firstParam instanceof ParameterizedType
+                        && CompletableFuture.class.equals(
+                                ((ParameterizedType) firstParam).getRawType())) {
+                    foundParam = true;
+                    genericParam = true;
+                }
+            }
+            if (!foundParam) {
+                throw new ValidationException(
+                        String.format(
+                                "Method '%s' of function class '%s' must have a first argument of type java.util.concurrent.CompletableFuture.",
+                                method.getName(), clazz.getName()));
+            }
+            if (!genericParam) {
```

Comment:
this we shouldn't validate because if someone uses a type inference with multiple types (i.e. no FunctionMappingExtractor), they might use raw types.

Reply:
Ah, that makes sense.  Ok removed.